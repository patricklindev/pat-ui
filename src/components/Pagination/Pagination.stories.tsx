import React from 'react';
import Pagination from './Pagination';



export default {
  title: 'Pagination',
  component: Pagination,
};

export const Filled = () => 
  (
    <div>
      <Pagination
        totalPages={5}
        pgnType={'filled'}
        pgnShape={'basic'}
        pgnColor={'primary'}
        pgnSize={'lg'}
      />
      <br/>
      <Pagination
        totalPages={8}
        pgnType={'filled'}
        pgnShape={'basic'}
        pgnColor={'secondary'}
        pgnSize={'lg'}
      />
      <br/>
      <Pagination
        totalPages={8}
        pgnType={'filled'}
        pgnShape={'rounded'}
        pgnColor={'primary'}
        pgnSize={'lg'}
      />
      <br/>
      <Pagination
        totalPages={8}
        pgnType={'filled'}
        pgnShape={'rounded'}
        pgnColor={'secondary'}
        pgnSize={'lg'}
      />
    </div>
    );

export const Outlined = () => 
  (
    <div>
      <Pagination
        totalPages={8}
        pgnType={'outlined'}
        pgnShape={'basic'}
        pgnColor={'primary'}
        pgnSize={'lg'}
      />
      <br/>
      <Pagination
        totalPages={8}
        pgnType={'outlined'}
        pgnShape={'basic'}
        pgnColor={'secondary'}
        pgnSize={'lg'}
      />
      <br/>
      <Pagination
        totalPages={8}
        pgnType={'outlined'}
        pgnShape={'rounded'}
        pgnColor={'primary'}
        pgnSize={'lg'}
      />
      <br/>
      <Pagination
        totalPages={8}
        pgnType={'outlined'}
        pgnShape={'rounded'}
        pgnColor={'secondary'}
        pgnSize={'lg'}
      />
    </div>
    );

