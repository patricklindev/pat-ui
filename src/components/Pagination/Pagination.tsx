import React, {FC, useState} from 'react';
import { classNames } from '../../utils/classNames';

export type paginationType = | 'filled' | 'outlined';
export type paginationshape = | 'basic'  | 'rounded';
export type paginationSize = 'lg' | 'md' | 'sm';
export type paginationColor = 'primary' | 'secondary' | 'default'

export interface IPaginationProps {
    className?:string;
    /** set Pagination shape */
    pgnShape?: paginationshape;
    /** set Pagination type */
    pgnType?: paginationType;
    /** set Pagination color */
    pgnColor?: paginationColor;
    /** set Pagination type */
    pgnSize?: paginationSize;
    /** total number of pages*/  
    totalPages: number;
}


export const Pagination: FC<IPaginationProps> = (props) => {
  const {totalPages, pgnShape, pgnType, pgnColor, pgnSize, className} = props
    let styleClasses = classNames('pgn', {
      [`pgn-${pgnType}`]: true,
        [`pgn-${pgnShape}`]: true,
        [`pgn-${pgnColor}`]: true,
        [`pgn-${pgnSize}`]: !!pgnSize,
      });

      if (className) {
        styleClasses += ' ' + className;
      } 

      let buttonId = classNames('btn')
    
      const [currentPage, setCurrentPage] = useState(1)
      const [btnActive, setBtnActive] = useState(0)
      const handleChange = (page:number) =>{
        setBtnActive(btnActive === 0 ? 1 : 0)
        setCurrentPage(page)
        console.log('The selected page is:', page)
    }
    return (

          <div className={styleClasses}>

           {currentPage !== 1 && (//Previous button will only appear when current page is !== 1
              <button
                onClick={() => handleChange(currentPage -1)}
                type="button"
                  className={buttonId}
              >
                &lt;
              </button>
            )}

            <button //first page button this will always be visible
              onClick={() => handleChange(1)}
              type="button"
              className={buttonId}
              id={currentPage === 1 ? '1': '0'}
            >
              {1}
            </button>

            {currentPage > 3 && //if selected currentPage is > 3 . . . will
            <div className={'elip'}>. . .</div>}
            
            {currentPage === totalPages && totalPages > 3 && (//selected page = last page and there are more than 3 currentPages then it will show the previous 2 pages
              <button
                onClick={() => handleChange(currentPage - 2)}
                type="button"
                className={buttonId}
              >
                {currentPage - 2}
              </button>
            )}

            {currentPage > 2 && (
              <button //in any case if current page is > 2 the button before current page will be shown
                onClick={() => handleChange(currentPage - 1)}
                type="button"
                className={buttonId}
              >
                {currentPage - 1}
              </button>
            )}

            {currentPage !== 1 && currentPage !== totalPages && (
              <button //current page will always show
                onClick={() => handleChange(currentPage)}
                type="button"
                className={buttonId}
                id={currentPage+''}
              >
                {currentPage}
              </button>
            )}
            {currentPage < totalPages - 1 && (
              <button //if current page is just before the last page last page will show
                onClick={() => handleChange(currentPage + 1)}
                type="button"
                className={buttonId}
              >
                {currentPage + 1}
              </button>
            )}
            {currentPage === 1 && totalPages > 3 && (
              <button //when in the first page it will show two more pages
                onClick={() => handleChange(currentPage + 2)}
                type="button"
                className={'btn'}
                
              >
                {currentPage + 2}
              </button>
            )}
            {currentPage < totalPages - 2 && <div //if page is less than the second page from the last elipses will appear 
            className={'elip'}>. . .</div>}

            {totalPages !== 1 && <button 
              onClick={() => handleChange(totalPages)}//last page button
              type="button"
              className={buttonId}
              id={currentPage === totalPages ? totalPages+'': '0'}
            >
              {totalPages}
            </button>}
            {currentPage !== totalPages && (//next page button will always appear unless current page is last page
              <button
                onClick={() => handleChange(currentPage + 1)}
                type="button"
                className={buttonId}
              >
                &gt;
              </button>
            )}

          </div>

      );
    };
export default Pagination;