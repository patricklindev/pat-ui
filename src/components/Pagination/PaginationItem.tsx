import React,{ FC } from 'react'
type liTypes = React.DetailedHTMLProps<React.HTMLAttributes<HTMLLIElement>, HTMLLIElement>
interface IPaginationItem { 
    callBack: ()=> void;
    builtClass?: string
} 

const PaginationItem: FC<IPaginationItem> = (props)=>{
    const { builtClass ='', callBack, children, ...rest} = props
    
    function onClickHandler(e:React.MouseEvent<HTMLDivElement>): void{ 
        callBack()
    }
    
    return(
        <div data-testid='pageButtonOuterHolder' role={'button'} className={builtClass} onClick={onClickHandler}>
        <li data-testid='pageButtons' {...(rest as liTypes)} >
           {children}
        </li>
        </div>
    )
}

export default PaginationItem
