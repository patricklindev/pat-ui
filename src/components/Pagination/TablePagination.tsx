import React from 'react';

import Button from '../Button/Button';

export type TablePaginationProps = {
  rowsPerPage: number;
  totalPageNumber: number;
  currentPage: number;
  onPageChange: Function;
};

export default function TablePagination({
  rowsPerPage,
  totalPageNumber,
  currentPage,
  onPageChange,
}: TablePaginationProps) {
  let endPage = rowsPerPage - 1 + currentPage;

  const onNext = () => {
    endPage = rowsPerPage + rowsPerPage;
    let newPage = currentPage + rowsPerPage;
    onPageChange(newPage);
  };

  const onPrevious = () => {
    endPage = endPage - rowsPerPage;
    let newPage = currentPage - rowsPerPage;
    onPageChange(newPage);
  };

  return (
    <div data-testid="tablePagination">
      <span>Rows per page: {rowsPerPage}</span>
      <div>
        {`${currentPage}-${endPage} of ${totalPageNumber}`}

        <Button disabled={currentPage === 1} onClick={onPrevious}>
          &#8249;
        </Button>
        {'      '}
        <Button disabled={endPage >= totalPageNumber} onClick={onNext}>
          &#8250;
        </Button>
      </div>
    </div>
  );
}
