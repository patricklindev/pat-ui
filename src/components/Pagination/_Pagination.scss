.pgn-outlined.pgn-primary {
    .btn[id] {
        background-color: #ffd8e6;
        border: 1px solid #F50057;
        color: #F50057;
        &:hover{
            background-color: $gray-400;
            border: none;
            color: rgba(0, 0, 0, 0.87);
        }
    }
    .btn[id^="0"], .btn{
        background-color: rgba(0, 0, 0, 0.08);
        border: none;
        color: rgba(0, 0, 0, 0.87);
        &:hover{
            background-color: $gray-400;
            border: none;
            color: black;
        }
    }
  }


  .pgn-outlined.pgn-secondary {
    .btn[id] {
        background-color: #dde2ff;
        border: 1px solid #3F51B5;
        color: #3F51B5;
        &:hover{
            background-color: $gray-400;
            border: none;
            color: rgba(0, 0, 0, 0.87);
        }
    }
    .btn[id^="0"], .btn{
        background-color: rgba(0, 0, 0, 0.08);
        border: none;
        color: rgba(0, 0, 0, 0.87);
        &:hover{
            background-color: $gray-400;
            border: none;
            color: rgba(0, 0, 0, 0.87);
        }
    }
  }


  .pgn-filled.pgn-primary{
    .btn[id] {
        border: none;
        background-color: #F50057;
        color: white;
        &:hover{
            background-color: $gray-400;
            border: none;
            color: rgba(0, 0, 0, 0.87);
        }
    }
    .btn[id^="0"], .btn{
        background-color: rgba(0, 0, 0, 0.08);
        border: none;
        color: rgba(0, 0, 0, 0.87);
        &:hover{
            background-color: $gray-400;
            border: none;
            color: rgba(0, 0, 0, 0.87);
        }
    }

}

.pgn-filled.pgn-secondary{
    .btn[id] {
        border: none;
        background-color: #3F51B5;
        color: white;
        &:hover{
            background-color: $gray-400;
            border: none;
            color: rgba(0, 0, 0, 0.87);
        }
    }
    .btn[id^="0"], .btn{
        background-color: rgba(0, 0, 0, 0.08);
        border: none;
        color: rgba(0, 0, 0, 0.87);
        &:hover{
            background-color: $gray-400;
            border: none;
            color: rgba(0, 0, 0, 0.87);
        }
    }
}


.pgn-basic{
    display: flex;
    flex-direction: row;
        .btn {
            border-radius: 100%;
            margin: 2px;
            display: flex;
            align-items: center;
            justify-content: center;
        }

}

.pgn-rounded{
    display: flex;
    flex-direction: row;
        .btn {
            border-radius: 10%;
            margin: 2px;
            align-items: center;
            display: flex;
            justify-content: center;
        }

}



.pgn-md {
    .btn{
        width: 32px;
        height: 32px;
    }
    .elip {
        display: flex;
        align-self: flex-end;
        font-size: 1.5rem;
    }
    .pgn-prev-next-btn{
        width: 32px;
        height: 32px;
        border: none;
    }
}


.pgn-lg {
    .btn{
        width: 40px;
        height: 40px;
    }
    .elip {
        display: flex;
        align-self: flex-end;
        font-size: 2rem;
    }
    .pgn-prev-next-btn{
        width: 40px;
        height: 40px;
        border: none;
    }
}

.pgn-sm {
    .btn{
        width: 26px;
        height: 26px;
    }
    .elip {
        display: flex;
        align-self: flex-end;
        font-size: 1rem;
    }
    .pgn-prev-next-btn{
        width: 26px;
        height: 26px;
        border: none;
    }
}
