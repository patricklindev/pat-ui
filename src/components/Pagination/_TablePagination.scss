.table-pagination {
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: $list-inline-padding * 3;

  .table-pagination__row__option {
    display: flex;

    .table-pagination__select {
      border: none;

      &:hover {
        cursor: pointer;
      }
    }
  }

  .pagination__icons__btn-icon {
    display: flex;
    justify-content: center;
    align-items: center;
    background: $white;
    border: none;
    outline: none;
    border-radius: $border-radius-sm;
  }

  .pagination__icons--disabled {
    fill: $gray-400;
  }

  .pagination__icons--default {
    &.pagination__icons__btn-icon {
      @include button-size(
        $btn-padding-y-default,
        $btn-padding-x-default,
        $btn-font-size-default,
        $border-radius-round
      );
    }
    &.pagination__icons__btn-arrow {
      width: $btn-padding-x-default * 4;
      height: $btn-padding-x-default * 4;
    }
    &:hover:not(.pagination__icons--disabled) {
      background-color: $gray-200;
    }
  }
}
