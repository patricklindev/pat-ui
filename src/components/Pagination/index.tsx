export { default as Pagination } from './Pagination';
export { default as TablePagination } from './TablePaginationContainer';
