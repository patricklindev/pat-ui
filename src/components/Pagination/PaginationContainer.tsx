//this component is for story book and test only
import React, {
  useState,
  ButtonHTMLAttributes,
  AnchorHTMLAttributes,
  FC,
} from 'react';

import Pagination from './Pagination';

export type PaginationSize = 'lg' | 'sm';
export type PaginationType =
  | 'primary'
  | 'secondary'
  | 'danger'
  | 'default'
  | 'link';

export interface IPaginationProps {
  /** set customized style */
  className?: string;
  /** set button size */
  pagSize?: PaginationSize;
  /** set button type */
  pagType?: PaginationType;
  /** set disabled button */
  disabled?: boolean;
  /** set pagination number */
  totalPageNumber?: number;
  currentPage?: number;
  rowsPerPage?: number;
  shape?: string;
  variant?: string;
}

type NativeButtonProps = IPaginationProps &
  ButtonHTMLAttributes<HTMLButtonElement>;
type NativeAchorButtonProps = IPaginationProps &
  AnchorHTMLAttributes<HTMLAnchorElement>;
export type PaginationProps = NativeButtonProps | NativeAchorButtonProps;

export const PaginationContainer: FC<PaginationProps> = ({
  pagType = 'default',
  pagSize,
  shape = 'rounded',
  disabled = false,
  variant = '',
}) => {
  const [page, setPage] = useState(1);

  const handleClickPage = (p: number) => {
    setPage(p);
  };

  return (
    <Pagination
      currentPage={page}
      onPageChange={handleClickPage}
      pagType={pagType}
      pagSize={pagSize}
      shape={shape}
      disabled={disabled}
      variant={variant}
    >
      Default Pagination
    </Pagination>
  );
};

export default PaginationContainer;
