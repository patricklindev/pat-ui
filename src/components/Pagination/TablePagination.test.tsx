import React from 'react';

import { IPaginationProps } from '../../utils/hooks/usePagination';
import Pagination from './Pagination';
import TablePagination from './TablePagination';

import { render, screen, fireEvent } from '@testing-library/react';

describe('TablePagination', () => {
  it('should match snapshot', () => {
    // default props
    const defaultProps = {
      count: 20,
    };
    const { asFragment } = render(<TablePagination {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render table pagination', () => {
    // enable table pagination
    const tableIPaginationProps: IPaginationProps = {
      count: 20,
      paginationType: 'table',
    };
    render(<Pagination {...tableIPaginationProps} />);
    const tablePaginationDiv = screen.getByTestId(
      'table-pagination'
    ) as HTMLElement;

    expect(tablePaginationDiv).toHaveClass('table-pagination');

    // default to disable previous button
    const prevBtn = screen.getByTestId('prev-btn');

    expect(prevBtn).toHaveClass('pagination__icons--disabled');
  });

  it('should disable next button', () => {
    // enable table pagination
    const tableIPaginationProps: IPaginationProps = {
      count: 20,
      paginationType: 'table',
      onPageChange: jest.fn(),
    };
    render(<Pagination {...tableIPaginationProps} />);
    const prevBtn = screen.getByTestId('prev-btn');
    const nextBtn = screen.getByTestId('next-btn');

    expect(prevBtn).toHaveClass('pagination__icons--disabled');
    expect(nextBtn).not.toHaveClass('pagination__icons--disabled');

    expect(tableIPaginationProps.onPageChange).toHaveBeenCalledTimes(0);
    fireEvent.click(nextBtn);
    expect(tableIPaginationProps.onPageChange).toHaveBeenCalledTimes(1);

    expect(prevBtn).not.toHaveClass('pagination__icons--disabled');
    expect(nextBtn).toHaveClass('pagination__icons--disabled');
  });
});
