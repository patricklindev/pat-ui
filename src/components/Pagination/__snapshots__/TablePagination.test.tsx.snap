// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TablePagination should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="table-pagination"
    data-testid="table-pagination"
  >
    <div>
      Rows per page:
    </div>
    <div
      aria-disabled="false"
      aria-label="table pagination"
      class="table-pagination__row__option"
    >
      <select
        class="table-pagination__select"
      >
        <option
          value="10"
        >
          10
        </option>
        <option
          value="25"
        >
          25
        </option>
        <option
          value="50"
        >
          50
        </option>
        <option
          value="100"
        >
          100
        </option>
      </select>
    </div>
    <div
      aria-label="current page"
    >
      <div>
        1 - 10 of 20 
      </div>
    </div>
    <div
      class="previous button"
    >
      <button
        aria-disabled="true"
        class="pagination__icons__btn-icon pagination__icons__btn-arrow pagination__icons--default pagination__icons--disabled "
        data-testid="prev-btn"
        disabled=""
      >
        <svg
          aria-hidden="true"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
      </button>
    </div>
    <div
      aria-label="next button"
    >
      <button
        aria-disabled="false"
        class="pagination__icons__btn-icon pagination__icons__btn-arrow pagination__icons--default  "
        data-testid="next-btn"
      >
        <svg
          aria-hidden="true"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
